File: lib/src/main/java/com/gerwalex/lib/main/BasicActivity.java
package com.gerwalex.lib.main;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.PreferenceManager;

/**
 * Basis-Activity  mit einem simplen Layout. Stellt ein Bundle args mit den extras aus Intent
 * bereit.
 */
public abstract class BasicActivity extends AppCompatActivity {
    public static final String ACTIONBARSUBTITLE = "ACTIONBARSUBTITLE";
    public static final String ACTIONBARTITLE = "ACTIONBARTITLE";
    /**
     * Bundle fuer Argumente. Wird in SaveStateInstance gesichert und in onCreate
     * wiederhergestellt.
     */
    protected final Bundle args = new Bundle();
    protected SharedPreferences prefs;
    private Handler handler;

    /**
     * Hides a Keyboard
     *
     * @see "stackoverflow.com/questions/1109022/close-hide-the-android-soft -keyboard"
     */
    @SuppressWarnings("unused")
    public void hide_keyboard() {
        View view = getCurrentFocus();
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    /**
     * Bereitstellen der Argumente aus Intent bzw savedStateInstance in args. Ist in den Args ein
     * String unter "NEXTACTIVIT" vorhanden, wird diese Activity nachgestartet Setzen der View
     *
     * @param savedInstanceState SavedState
     */
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        prefs = PreferenceManager.getDefaultSharedPreferences(this);
        if (savedInstanceState != null) {
            args.putAll(savedInstanceState);
        } else {
            Bundle extras = getIntent().getExtras();
            if (extras != null) {
                args.putAll(extras);
            }
        }
    }

    /**
     * Setzt den Subtitle aus args neu
     */
    @Override
    protected void onResume() {
        super.onResume();
        setSubTitle(args.getString(ACTIONBARSUBTITLE));
    }

    /**
     * Sicherung aller Argumente
     */
    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putAll(args);
        super.onSaveInstanceState(outState);
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    /**
     * Startet ein Runnable verzoegert auf dem MainThread
     *
     * @param r           Runnable
     * @param delayMillis Verzoegerung in Millis
     */
    public void postOnUiThread(Runnable r, long delayMillis) {
        if (handler == null) {
            handler = new Handler(getMainLooper());
        }
        handler.postDelayed(r, delayMillis);
    }

    /**
     * Setzt den SubTitle in der SupportActionBar
     *
     * @param subTitleResID resID des Subtitles
     */
    public final void setSubTitle(@StringRes int subTitleResID) {
        setSubTitle(getString(subTitleResID));
    }

    /**
     * Setzt den SubTitle in der SupportActionBar, wenn vorhanden
     *
     * @param subTitle Text des Subtitles
     */
    public final void setSubTitle(CharSequence subTitle) {
        ActionBar bar = getSupportActionBar();
        if (bar != null) {
            bar.setSubtitle(subTitle);
        }
        args.putCharSequence(ACTIONBARSUBTITLE, subTitle);
    }

    @Override
    public void setTitle(CharSequence title) {
        super.setTitle(title);
    }

    @Override
    public void setTitle(int titleId) {
        super.setTitle(titleId);
    }
}

File: lib/src/main/java/com/gerwalex/lib/main/BasicActivity.kt
package com.gerwalex.lib.main

import android.content.SharedPreferences
import android.os.Bundle
import android.os.Handler
import android.view.inputmethod.InputMethodManager
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.preference.PreferenceManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlin.coroutines.CoroutineContext

/**
 * Basis-Activity  mit einem simplen Layout. Stellt ein Bundle args mit den extras aus Intent
 * bereit.
 */
abstract class BasicActivity : AppCompatActivity(), CoroutineScope {
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job
    private lateinit var job: Job

    /**
     * Bundle fuer Argumente. Wird in SaveStateInstance gesichert und in onCreate
     * wiederhergestellt.
     */
    protected val args = Bundle()
    protected lateinit var prefs: SharedPreferences
    private lateinit var handler: Handler

    /**
     * Hides a Keyboard
     *
     * @see "stackoverflow.com/questions/1109022/close-hide-the-android-soft -keyboard"
     */
    fun hide_keyboard() {
        val view = currentFocus
        if (view != null) {
            val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    /**
     * Cancelt eventuell laufende Jobs
     */
    override fun onDestroy() {
        super.onDestroy()
        job.cancel()
    }

    /**
     * Bereitstellen der Argumente aus Intent bzw savedStateInstance in args. Ist in den Args ein
     * String unter "NEXTACTIVIT" vorhanden, wird diese Activity nachgestartet Setzen der View
     *
     *
     * @param savedInstanceState SavedState
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        job = Job()
        prefs = PreferenceManager.getDefaultSharedPreferences(this)
        if (savedInstanceState != null) {
            args.putAll(savedInstanceState)
        } else {
            val extras = intent
                .extras
            if (extras != null) {
                args.putAll(extras)
            }
        }
    }

    /**
     * Setzt den Subtitle aus args neu
     */
    override fun onResume() {
        super.onResume()
        setSubTitle(args.getString(ACTIONBARSUBTITLE))
    }

    /**
     * Sicherung aller Argumente
     */
    override fun onSaveInstanceState(outState: Bundle) {
        outState.putAll(args)
        super.onSaveInstanceState(outState)
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    /**
     * Startet ein Runnable verzoegert auf dem MainThread
     *
     * @param r           Runnable
     * @param delayMillis Verzoegerung in Millis
     */
    fun postOnUiThread(r: Runnable, delayMillis: Long) {
        handler = Handler(mainLooper)
        handler.postDelayed(r, delayMillis)
    }

    /**
     * Setzt den SubTitle in der SupportActionBar
     *
     * @param subTitleResID resID des Subtitles
     */
    fun setSubTitle(@StringRes subTitleResID: Int) {
        setSubTitle(getString(subTitleResID))
    }

    /**
     * Setzt den SubTitle in der SupportActionBar, wenn vorhanden
     *
     * @param subTitle Text des Subtitles
     */
    fun setSubTitle(subTitle: CharSequence?) {
        actionBar?.let {
            it.subtitle = subTitle
        }
        args.putCharSequence(ACTIONBARSUBTITLE, subTitle)
    }

    override fun setTitle(title: CharSequence) {
        super.setTitle(title)
    }

    companion object {
        const val ACTIONBARSUBTITLE = "ACTIONBARSUBTITLE"
        const val ACTIONBARTITLE = "ACTIONBARTITLE"
    }
}